File: portfolio-project/frontend/src/lib/dossier.ts
import type { Locale } from "@/content/site";
import type { ProjectDossier } from "@/services/types";
import type { Project } from "@/services/types";

export type C4NodeKind =
  | "person"
  | "system"
  | "client"
  | "container"
  | "component"
  | "store"
  | "queue"
  | "external";

export interface C4Node {
  kind: C4NodeKind;
  title: string;
  sub?: string;
  leaf?: boolean;
}

export interface C4Level {
  label: string;
  note?: string;
  tiers: C4Node[][];
}

export interface Adr {
  id: string;
  title: string;
  status: string;
  date?: string;
  context: string;
  decision: string;
  tradeoff?: string;
}

export interface LogEntry {
  hash: string;
  tag?: string;
  date: string;
  title: string;
  note?: string;
}

export interface SequenceMessage {
  from: string;
  to: string;
  label: string;
  kind?: "return";
}

export interface SequenceData {
  actors: string[];
  messages: SequenceMessage[];
}

export interface Entity {
  name: string;
  kind?: "table" | "class" | "abstract" | "interface" | "enum";
  rows?: string[];
}

export interface SchemaRelation {
  from: string;
  label: string;
  to: string;
}

export interface SchemaData {
  tiers: Entity[][];
  relations?: SchemaRelation[];
}

export type FlowKind =
  | "start"
  | "end"
  | "state"
  | "final"
  | "step"
  | "decision"
  | "error"
  | "store";

export interface FlowNode {
  kind: FlowKind;
  title: string;
  sub?: string;
  via?: string;
}

export interface TiersData {
  tiers: FlowNode[][];
  notes?: string[];
}

export interface MatrixRow {
  label: string;
  cells: string[];
}

export interface MatrixData {
  cols: string[];
  rows: MatrixRow[];
}

export type Diagram =
  | { id: string; kind: "c4"; title: string; note?: string; data: C4Level[] }
  | { id: string; kind: "sequence"; title: string; note?: string; data: SequenceData }
  | { id: string; kind: "schema"; title: string; note?: string; data: SchemaData }
  | { id: string; kind: "tiers"; title: string; note?: string; data: TiersData }
  | { id: string; kind: "matrix"; title: string; note?: string; data: MatrixData };

export interface DossierMetric {
  value: string;
  label: string;
  note?: string;
}

export interface GalleryItem {
  id: string;
  src: string;
  caption: string;
  hint?: string;
}

export interface ProjectDetail {
  metrics: DossierMetric[];
  c4: C4Level[];
  adrs: Adr[];
  log: LogEntry[];
  diagrams: Diagram[];
  gallery: GalleryItem[];
}

export interface DossierProject {
  slug: string;
  title: string;
  summary: string;
  description: string;
  impact: string;
  technologies: string[];
  featured: boolean;
  details?: ProjectDetail;
}

interface ApiC4Node {
  kind: C4NodeKind;
  title: string;
  sub?: string | null;
  leaf?: boolean;
}

interface ApiC4Level {
  label: string;
  note?: string | null;
  tiers: ApiC4Node[][];
}

const byOrder = <T extends { display_order: number }>(items: T[]): T[] =>
  [...items].sort((a, b) => a.display_order - b.display_order);

const mapC4Level = (level: ApiC4Level): C4Level => ({
  label: level.label,
  note: level.note ?? undefined,
  tiers: level.tiers.map((tier) =>
    tier.map((node) => ({
      kind: node.kind,
      title: node.title,
      sub: node.sub ?? undefined,
      leaf: node.leaf,
    })),
  ),
});

const mapDiagram = (diagram: ProjectDossier["diagrams"][number]): Diagram => {
  const data = diagram.data as unknown as Record<string, unknown>;
  const note = diagram.note ?? undefined;

  if (diagram.kind === "c4") {
    return {
      id: diagram.id,
      kind: "c4",
      title: diagram.title,
      note,
      data: ((data.levels as ApiC4Level[] | undefined) ?? []).map(mapC4Level),
    };
  }

  if (diagram.kind === "sequence") {
    const messages = Array.isArray(data.messages) ? data.messages : [];
    return {
      id: diagram.id,
      kind: "sequence",
      title: diagram.title,
      note,
      data: {
        actors: Array.isArray(data.actors) ? (data.actors as string[]) : [],
        messages: messages.map((message) => {
          const item = message as { from: string; to: string; label: string; kind?: string | null };
          return {
            from: item.from,
            to: item.to,
            label: item.label,
            kind: item.kind === "return" ? "return" : undefined,
          };
        }),
      },
    };
  }

  if (diagram.kind === "schema") {
    const tiers = Array.isArray(data.tiers) ? data.tiers : [];
    return {
      id: diagram.id,
      kind: "schema",
      title: diagram.title,
      note,
      data: {
        tiers: tiers as Entity[][],
        relations: Array.isArray(data.relations) ? (data.relations as SchemaRelation[]) : [],
      },
    };
  }

  if (diagram.kind === "matrix") {
    return {
      id: diagram.id,
      kind: "matrix",
      title: diagram.title,
      note,
      data: {
        cols: Array.isArray(data.cols) ? (data.cols as string[]) : [],
        rows: Array.isArray(data.rows) ? (data.rows as MatrixRow[]) : [],
      },
    };
  }

  return {
    id: diagram.id,
    kind: "tiers",
    title: diagram.title,
    note,
    data: {
      tiers: (Array.isArray(data.tiers) ? data.tiers : []) as FlowNode[][],
      notes: Array.isArray(data.notes) ? (data.notes as string[]) : [],
    },
  };
};

export function toDossierProject(
  project: Project,
  dossier: ProjectDossier | null,
  _locale: Locale,
): DossierProject {
  const details: ProjectDetail | undefined = dossier
    ? {
        metrics: byOrder(dossier.metrics).map((metric) => ({
          value: metric.value,
          label: metric.label,
          note: metric.note ?? undefined,
        })),
        c4: byOrder(dossier.c4).map(mapC4Level),
        adrs: byOrder(dossier.adrs).map((adr) => ({
          id: adr.id,
          title: adr.title,
          status: adr.status,
          date: adr.date ?? undefined,
          context: adr.context,
          decision: adr.decision,
          tradeoff: adr.tradeoff ?? undefined,
        })),
        log: byOrder(dossier.log).map((entry) => ({
          hash: entry.hash,
          tag: entry.tag ?? undefined,
          date: entry.date,
          title: entry.title,
          note: entry.note ?? undefined,
        })),
        diagrams: byOrder(dossier.diagrams).map(mapDiagram),
        gallery: byOrder(dossier.gallery).map((item) => ({
          id: item.id,
          src: item.src,
          caption: item.caption,
          hint: item.hint ?? undefined,
        })),
      }
    : undefined;

  if (details && details.c4.length > 0) {
    details.diagrams = [
      {
        id: "c4",
        kind: "c4",
        title: "C4 Model",
        note: "semantic zoom — click nodes to descend",
        data: details.c4,
      },
      ...details.diagrams,
    ];
  }

  return {
    slug: project.slug,
    title: project.title,
    summary: project.short_description?.trim() || project.description,
    description: project.description,
    impact: dossier?.impact ?? "",
    technologies: (project.technologies || []).map((technology) => technology.name),
    featured: project.featured,
    details,
  };
}

File: portfolio-project/backend/app/schemas/dossier.py
"""Typed request and response schemas for project dossiers."""

from decimal import Decimal
from typing import Annotated, Literal, Union
from urllib.parse import urlparse
import uuid

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class DossierSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class DossierMetricCreate(DossierSchema):
    value: str = Field(..., min_length=1, max_length=255)
    numeric_value: Decimal | None = Field(default=None, ge=0)
    label: str = Field(..., min_length=1, max_length=255)
    note: str | None = Field(default=None, max_length=2_000)
    display_order: int = Field(default=0, ge=0)


class DossierMetricResponse(DossierMetricCreate):
    id: uuid.UUID


C4NodeKind = Literal[
    "person",
    "system",
    "client",
    "container",
    "component",
    "store",
    "queue",
    "external",
]


class DossierC4NodeCreate(DossierSchema):
    kind: C4NodeKind
    title: str = Field(..., min_length=1, max_length=255)
    sub: str | None = Field(default=None, max_length=2_000)
    leaf: bool = False
    tier_order: int = Field(default=0, ge=0)
    display_order: int = Field(default=0, ge=0)


class DossierC4NodeResponse(DossierC4NodeCreate):
    id: uuid.UUID


class DossierC4LevelCreate(DossierSchema):
    label: str = Field(..., min_length=1, max_length=255)
    note: str | None = Field(default=None, max_length=2_000)
    tiers: list[list[DossierC4NodeCreate]] = Field(..., min_length=1, max_length=20)
    display_order: int = Field(default=0, ge=0)

    @model_validator(mode="after")
    def validate_tiers(self):
        if any(not tier for tier in self.tiers):
            raise ValueError("C4 tiers must contain at least one node")
        return self


class DossierC4LevelResponse(DossierSchema):
    id: uuid.UUID
    label: str
    note: str | None = None
    tiers: list[list[DossierC4NodeResponse]]
    display_order: int


class DossierAdrCreate(DossierSchema):
    id: str = Field(..., min_length=1, max_length=64)
    title: str = Field(..., min_length=1, max_length=255)
    status: str = Field(..., min_length=1, max_length=64)
    date: str | None = Field(default=None, max_length=32)
    context: str = Field(..., min_length=1, max_length=10_000)
    decision: str = Field(..., min_length=1, max_length=10_000)
    tradeoff: str | None = Field(default=None, max_length=10_000)
    display_order: int = Field(default=0, ge=0)


class DossierAdrResponse(DossierAdrCreate):
    row_id: uuid.UUID = Field(alias="id")
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class DossierLogEntryCreate(DossierSchema):
    hash: str = Field(..., min_length=1, max_length=64)
    tag: str | None = Field(default=None, max_length=64)
    date: str = Field(..., min_length=1, max_length=32)
    title: str = Field(..., min_length=1, max_length=255)
    note: str | None = Field(default=None, max_length=10_000)
    display_order: int = Field(default=0, ge=0)


class DossierLogEntryResponse(DossierLogEntryCreate):
    row_id: uuid.UUID = Field(alias="id")
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class SequenceMessage(DossierSchema):
    from_: str = Field(alias="from", min_length=1, max_length=255)
    to: str = Field(..., min_length=1, max_length=255)
    label: str = Field(..., min_length=1, max_length=2_000)
    kind: Literal["return"] | None = None
    model_config = ConfigDict(populate_by_name=True)


class C4DiagramData(DossierSchema):
    kind: Literal["c4"]
    levels: list[DossierC4LevelCreate] = Field(default_factory=list, max_length=8)


class SequenceDiagramData(DossierSchema):
    kind: Literal["sequence"]
    actors: list[str] = Field(..., min_length=1, max_length=50)
    messages: list[SequenceMessage] = Field(default_factory=list, max_length=200)


class Entity(DossierSchema):
    name: str = Field(..., min_length=1, max_length=255)
    kind: Literal["table", "class", "abstract", "interface", "enum"] | None = None
    rows: list[str] = Field(default_factory=list, max_length=100)


class SchemaRelation(DossierSchema):
    from_: str = Field(alias="from", min_length=1, max_length=255)
    label: str = Field(..., min_length=1, max_length=64)
    to: str = Field(..., min_length=1, max_length=255)
    model_config = ConfigDict(populate_by_name=True)


class SchemaDiagramData(DossierSchema):
    kind: Literal["schema"]
    tiers: list[list[Entity]] = Field(..., min_length=1, max_length=20)
    relations: list[SchemaRelation] = Field(default_factory=list, max_length=100)


FlowKind = Literal["start", "end", "state", "final", "step", "decision", "error", "store"]


class FlowNode(DossierSchema):
    kind: FlowKind
    title: str = Field(..., min_length=1, max_length=255)
    sub: str | None = Field(default=None, max_length=2_000)
    via: str | None = Field(default=None, max_length=255)


class TiersDiagramData(DossierSchema):
    kind: Literal["tiers"]
    tiers: list[list[FlowNode]] = Field(..., min_length=1, max_length=30)
    notes: list[str] = Field(default_factory=list, max_length=100)


class MatrixRow(DossierSchema):
    label: str = Field(..., min_length=1, max_length=255)
    cells: list[str] = Field(..., min_length=1, max_length=100)


class MatrixDiagramData(DossierSchema):
    kind: Literal["matrix"]
    cols: list[str] = Field(..., min_length=1, max_length=100)
    rows: list[MatrixRow] = Field(..., min_length=1, max_length=100)


DiagramData = Annotated[
    Union[
        C4DiagramData,
        SequenceDiagramData,
        SchemaDiagramData,
        TiersDiagramData,
        MatrixDiagramData,
    ],
    Field(discriminator="kind"),
]


class DossierDiagramCreate(DossierSchema):
    id: str = Field(..., min_length=1, max_length=128)
    kind: Literal["c4", "sequence", "schema", "tiers", "matrix"]
    title: str = Field(..., min_length=1, max_length=255)
    note: str | None = Field(default=None, max_length=2_000)
    data: DiagramData
    display_order: int = Field(default=0, ge=0)

    @model_validator(mode="after")
    def match_kind(self):
        if self.data.kind != self.kind:
            raise ValueError("diagram kind must match diagram data kind")
        return self


class DossierDiagramResponse(DossierDiagramCreate):
    row_id: uuid.UUID = Field(alias="id")
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class DossierGalleryItemCreate(DossierSchema):
    id: str = Field(..., min_length=1, max_length=128)
    src: str = Field(..., min_length=2, max_length=500)
    caption: str = Field(..., min_length=1, max_length=2_000)
    hint: str | None = Field(default=None, max_length=2_000)
    display_order: int = Field(default=0, ge=0)

    @field_validator("src")
    @classmethod
    def validate_src(cls, value: str) -> str:
        parsed = urlparse(value)
        if value.startswith("/") and not value.startswith("//"):
            return value
        if parsed.scheme in {"http", "https"} and parsed.netloc:
            return value
        raise ValueError("gallery source must be a site-relative or http(s) URL")


class DossierGalleryItemResponse(DossierGalleryItemCreate):
    row_id: uuid.UUID = Field(alias="id")
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class ProjectDossierUpsert(DossierSchema):
    impact_en: str = Field(..., min_length=1, max_length=10_000)
    impact_tr: str = Field(..., min_length=1, max_length=10_000)
    metrics: list[DossierMetricCreate] = Field(default_factory=list, max_length=20)
    c4: list[DossierC4LevelCreate] = Field(default_factory=list, max_length=8)
    adrs: list[DossierAdrCreate] = Field(default_factory=list, max_length=50)
    log: list[DossierLogEntryCreate] = Field(default_factory=list, max_length=100)
    diagrams: list[DossierDiagramCreate] = Field(default_factory=list, max_length=30)
    gallery: list[DossierGalleryItemCreate] = Field(default_factory=list, max_length=50)

    @model_validator(mode="after")
    def validate_unique_ids(self):
        for collection_name in ("adrs", "log", "diagrams", "gallery"):
            collection = getattr(self, collection_name)
            ids = [item.id if collection_name != "log" else item.hash for item in collection]
            if len(ids) != len(set(ids)):
                raise ValueError(f"duplicate identifiers in {collection_name}")
        return self


class ProjectDossierResponse(DossierSchema):
    id: uuid.UUID
    project_id: uuid.UUID
    project_slug: str
    impact: str
    metrics: list[DossierMetricResponse]
    c4: list[DossierC4LevelResponse]
    adrs: list[DossierAdrResponse]
    log: list[DossierLogEntryResponse]
    diagrams: list[DossierDiagramResponse]
    gallery: list[DossierGalleryItemResponse]


class AdminProjectDossierResponse(ProjectDossierResponse):
    impact_en: str
    impact_tr: str
